src/app/blog/blake3-hazmat-api/page.mdx

```diff
@@ -0,0 +1,156 @@
+import { BlogPostLayout } from '@/components/BlogPostLayout'
+import {ThemeImage} from '@/components/ThemeImage'
+
+export const post = {
+  draft: false,
+  author: 'Rüdiger Klaehn',
+  date: '2025-04-08',
+  title: 'The new BLAKE3 hazmat API',
+  description:
+    "The new BLAKE3 hazmat API, or why we can retire our BLAKE3 fork",
+}
+
+export const metadata = {
+    title: post.title,
+    description: post.description,
+    openGraph: {
+      title: post.title,
+      description: post.description,
+      images: [{
+        url: `/api/og?title=Blog&subtitle=${post.title}`,
+        width: 1200,
+        height: 630,
+        alt: post.title,
+        type: 'image/png',
+      }],
+      type: 'article'
+    }
+}
+
+export default (props) => <BlogPostLayout article={post} {...props} />
+
+In [iroh-blobs], we are using the [BLAKE3] hash function for content-addressing. To allow verified streaming, we need to access some internals of the hash function. For the last 18 months we have maintaned a [fork] of the BLAKE3 crate to allow us to access some internals of the hash function. With the latest release of BLAKE3, there is a new hazmat API that allows us to retire this fork.
+
+This blog post goes into a lot of detail on the previous guts API, the new hazmat API, and how we use these APIs in blobs. It will be of interest to other advanced users of the BLAKE3 hash function.
+
+# Short summary of BLAKE3
+
+The BLAKE3 hash function is a *tree* hash function. Briefly, the data gets divided into 1024 byte chunks, and each chunk is then hashed individually.
+
+To compute a root hash, pairs of hashes are combined with a binary combine function until just a single hash remains. To provide [domain separation], ***only*** the final root hash gets computed with a special flag is_root set to true.
+
+<div className="not-prose">
+<ThemeImage
+    lightSrc='/blog/blake3-hazmat-api/blake-tree-6.svg'
+    darkSrc='/blog/blake3-hazmat-api/blake-tree-6.svg'
+    alt='BLAKE3 forks'
+    width={800}
+    height={600}
+  />
+</div>
+
+# Why does iroh-blobs need internals access
+
+The iroh-blobs protocol allows for verified streaming of data, just like the [bao] crate. In order to do so we need to be able to compute *intermediate results* of the computation described above.
+
+<div className="not-prose">
+<ThemeImage
+    lightSrc='/blog/blake3-hazmat-api/blake-tree-21.svg'
+    darkSrc='/blog/blake3-hazmat-api/blake-tree-21.svg'
+    alt='BLAKE3 forks'
+    width={800}
+    height={600}
+  />
+</div>
+
+<Note>
+Most uses are only interested in the root hash like most users, and won't *ever* have to deal with intermediate results.
```

Comment:
"Most use cases are only interested in the root hash and won't *ever* have to deal with intermediate results." ?

"Most users are only interested in the root hash and won't *ever* have to deal with intermediate results." ?

Reply:
You could argue that use cases aren't interested in anything...

Reply:
So users?